airflow/metrics/otel_logger.py

```diff
@@ -0,0 +1,261 @@
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from __future__ import annotations
+
+import logging
+import random
+import warnings
+from typing import Callable
+
+from opentelemetry import metrics
+from opentelemetry.exporter.otlp.proto.http.metric_exporter import OTLPMetricExporter
+from opentelemetry.sdk.metrics import MeterProvider
+from opentelemetry.sdk.metrics._internal.export import ConsoleMetricExporter, PeriodicExportingMetricReader
+from opentelemetry.sdk.resources import SERVICE_NAME, Resource
+from opentelemetry.util.types import Attributes
+
+from airflow.configuration import conf
+from airflow.metrics.protocols import DeltaType, Timer, TimerProtocol
+from airflow.metrics.validators import AllowListValidator, validate_stat
+
+log = logging.getLogger(__name__)
+
+
+# This is currently the only UDC used.  If more are added, we should add a better system for this.
+UP_DOWN_COUNTERS = {"airflow.dag_processing.processes"}
+OTEL_NAME_MAX_LENGTH = 63
+METRIC_NAME_PREFIX = "airflow."
+
+
+def _is_up_down_counter(name):
+    return name in UP_DOWN_COUNTERS
+
+
+def _generate_key_name(name: str, attributes: Attributes = None):
+    return f"{name}{'_' + str(attributes) if attributes else ''}"
+
+
+class SafeOtelLogger:
+    """Otel Logger"""
+
+    def __init__(self, otel_provider, prefix: str = "airflow", allow_list_validator=AllowListValidator()):
+        self.otel: Callable = otel_provider
+        self.prefix: str = prefix
+        self.metrics_validator = allow_list_validator
+        self.meter = otel_provider.get_meter(__name__)
+        self.metrics_map = MetricsMap(self.meter)
+
+    @validate_stat
+    def incr(
+        self,
+        stat: str,
+        count: int = 1,
+        rate: float = 1,
+        tags: Attributes = None,
+        *,
+        back_compat_name: str = "",
+    ):
+        """
+        Increment stat by count.
+
+        :param stat: The name of the stat to increment.
+        :param count: A positive integer to add to the current value of stat.
+        :param rate: value between 0 and 1 that represents the sampled rate at
+            which the metric is going to be emitted.
+        :param tags: Tags to append to the stat.
+        :param back_compat_name: If possible, the metric will also be emitted
+            under this name for backward compatibility.
+        """
+        if (count < 0) or (rate < 0):
+            raise ValueError("count and rate must both be positive values.")
+        if rate < 1 and random.random() > rate:
+            return
+
+        if back_compat_name and self.metrics_validator.test(back_compat_name):
```

Comment:
Does the metrics_validator do the assertion of the name length for otel's short naming? Otherwise it's going to blow up when it hits the hotel code.

Reply:
No, I think I need to truncate it on the following line(89).  I'll also add a unit test which passes in a back_compat_name that is too long and makes sure it doesn't blow up.

Reply:
This parameter has been reverted/removed, so no longer relevant. Resolving.